Fix sequence numbers in profiler output

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42565 Fix sequence numbers in profiler output**

Summary:
After recent changes to the record function we record more
ranges in profiler output and also keep emitting sequence numbers for
all ranges.

Sequence numbers are used by external tools to correlate forward
and autograd ranges and with many ranges having the same sequence number
it becomes impossible to do this.

This PR ensures that we set sequence numbers only for the top-level
ranges and only in case when autograd is enabled.

Test Plan:
nvprof -fo trace.nvvp --profile-from-start off python test_script.py
test_script
https://gist.github.com/ilia-cher/2baffdd98951ee2a5f2da56a04fe15d0
then examining ranges in nvvp

Differential Revision: [D22938828](https://our.internmc.facebook.com/intern/diff/D22938828)

Commit:
Fix sequence numbers in profiler output

Summary:
After recent changes to the record function we record more
ranges in profiler output and also keep emitting sequence numbers for
all ranges.

Sequence numbers are used by external tools to correlate forward
and autograd ranges and with many ranges having the same sequence number
it becomes impossible to do this.

This PR ensures that we set sequence numbers only for the top-level
ranges and only in case when autograd is enabled.

Test Plan:
nvprof -fo trace.nvvp --profile-from-start off python test_script.py
test_script
https://gist.github.com/ilia-cher/2baffdd98951ee2a5f2da56a04fe15d0
then examining ranges in nvvp

Reviewers: ngimel

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

mcarilli commented:
Should close https://github.com/pytorch/pytorch/issues/42564.  By pure coincidence, the issue number is off by one from the PR number, which, considering the subject, proves the universe has a sense of humor.

Commit:
Update on "Fix sequence numbers in profiler output"


Summary:
After recent changes to the record function we record more
ranges in profiler output and also keep emitting sequence numbers for
all ranges.

Sequence numbers are used by external tools to correlate forward
and autograd ranges and with many ranges having the same sequence number
it becomes impossible to do this.

This PR ensures that we set sequence numbers only for the top-level
ranges and only in case when autograd is enabled.

Test Plan:
nvprof -fo trace.nvvp --profile-from-start off python test_script.py
test_script
https://gist.github.com/ilia-cher/2baffdd98951ee2a5f2da56a04fe15d0
then examining ranges in nvvp

Differential Revision: [D22938828](https://our.internmc.facebook.com/intern/diff/D22938828)

[ghstack-poisoned]

ilia-cher commented:
added test ( @ngimel )

Commit:
Update on "Fix sequence numbers in profiler output"


Summary:
After recent changes to the record function we record more
ranges in profiler output and also keep emitting sequence numbers for
all ranges.

Sequence numbers are used by external tools to correlate forward
and autograd ranges and with many ranges having the same sequence number
it becomes impossible to do this.

This PR ensures that we set sequence numbers only for the top-level
ranges and only in case when autograd is enabled.

Test Plan:
nvprof -fo trace.nvvp --profile-from-start off python test_script.py
test_script
https://gist.github.com/ilia-cher/2baffdd98951ee2a5f2da56a04fe15d0
then examining ranges in nvvp

Differential Revision: [D22938828](https://our.internmc.facebook.com/intern/diff/D22938828)

[ghstack-poisoned]

Commit:
Update on "Fix sequence numbers in profiler output"


Summary:
After recent changes to the record function we record more
ranges in profiler output and also keep emitting sequence numbers for
all ranges.

Sequence numbers are used by external tools to correlate forward
and autograd ranges and with many ranges having the same sequence number
it becomes impossible to do this.

This PR ensures that we set sequence numbers only for the top-level
ranges and only in case when autograd is enabled.

Test Plan:
nvprof -fo trace.nvvp --profile-from-start off python test_script.py
test_script
https://gist.github.com/ilia-cher/2baffdd98951ee2a5f2da56a04fe15d0
then examining ranges in nvvp

Differential Revision: [D22938828](https://our.internmc.facebook.com/intern/diff/D22938828)

[ghstack-poisoned]

Commit:
Update on "Fix sequence numbers in profiler output"


Summary:
After recent changes to the record function we record more
ranges in profiler output and also keep emitting sequence numbers for
all ranges.

Sequence numbers are used by external tools to correlate forward
and autograd ranges and with many ranges having the same sequence number
it becomes impossible to do this.

This PR ensures that we set sequence numbers only for the top-level
ranges and only in case when autograd is enabled.

Test Plan:
nvprof -fo trace.nvvp --profile-from-start off python test_script.py
test_script
https://gist.github.com/ilia-cher/2baffdd98951ee2a5f2da56a04fe15d0
then examining ranges in nvvp

Differential Revision: [D22938828](https://our.internmc.facebook.com/intern/diff/D22938828)

[ghstack-poisoned]

dzhulgakov reviewed (commented):
Looks good, will let other folks to take a look too.

Do we have first-class support for folding backwards into aggregation btw? I think that'd be useful as an option in `profile()`'s output

ngimel commented on test/test_autograd.py:
```diff
@@ -2705,6 +2705,52 @@ def test_profiler(self):
+        found_add = found_sum = False
+        found_bwd_add = found_bwd_sum = False
+        found_empty = False
+        for e in p.function_events:
```
I assume this is after filtering, so you won't have multiple function events with the same name?

ilia-cher replied:
yes, this issue with double logging is not visible here, because its after filtering;
I also err'd on the side of failing a test, so added extra checks.

ngimel commented on torch/csrc/autograd/function.h:
```diff
@@ -91,7 +91,7 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
   /// higher sequence numbers prioritized before lower sequence numbers.
   explicit Node(
-      uint64_t sequence_nr,
+      int64_t sequence_nr,
```
why did you need int64_t here? Normally autograd should not get negative seq_nrs?

ilia-cher replied:
it's not necessary here, will upd

ngimel commented:
> Do we have first-class support for folding backwards into aggregation btw? I think that'd be useful as an option in `profile()`'s output

Not yet, but having sequence_nr as part of function_event is a first step towards that (before it was available only via nvtx labels, so only pyprof was using it)

Commit:
Update on "Fix sequence numbers in profiler output"


Summary:
After recent changes to the record function we record more
ranges in profiler output and also keep emitting sequence numbers for
all ranges.

Sequence numbers are used by external tools to correlate forward
and autograd ranges and with many ranges having the same sequence number
it becomes impossible to do this.

This PR ensures that we set sequence numbers only for the top-level
ranges and only in case when autograd is enabled.

Test Plan:
nvprof -fo trace.nvvp --profile-from-start off python test_script.py
test_script
https://gist.github.com/ilia-cher/2baffdd98951ee2a5f2da56a04fe15d0
then examining ranges in nvvp

Differential Revision: [D22938828](https://our.internmc.facebook.com/intern/diff/D22938828)

[ghstack-poisoned]

Commit:
Update on "Fix sequence numbers in profiler output"


Summary:
After recent changes to the record function we record more
ranges in profiler output and also keep emitting sequence numbers for
all ranges.

Sequence numbers are used by external tools to correlate forward
and autograd ranges and with many ranges having the same sequence number
it becomes impossible to do this.

This PR ensures that we set sequence numbers only for the top-level
ranges and only in case when autograd is enabled.

Test Plan:
nvprof -fo trace.nvvp --profile-from-start off python test_script.py
test_script
https://gist.github.com/ilia-cher/2baffdd98951ee2a5f2da56a04fe15d0
then examining ranges in nvvp

Differential Revision: [D22938828](https://our.internmc.facebook.com/intern/diff/D22938828)

[ghstack-poisoned]

ilia-cher commented:
comment and lint
